source/application/GUI/Core/DlgBase.h

```diff
@@ -0,0 +1,138 @@
+ * success. When creating a Prompt dialog, SpawnDialogBox() will return the result of the dialog
+ * box when it exits.
+ *
+ * The purpose of the overloaded WinAPI calls are to send messages to the window and it's controls,
+ * without needing to know the HWND.
+ * If you provide an external interface to your class, don't just proxy all paramterized calls,
+ * wrap them where it makes sense.
+ * You should have class members that look like this (example code, do not actually use):
+ * bool SendCloseMessage()
+ * {
+ *    SendMessage(WM_CLOSE, 0, 0);
+ *    return true;
+ * }
+ * These overloaded SendMessage and SendDlgItemMessage functions can now fail with LRESULT -1 and
+ * ERROR_CAN_NOT_COMPLETE, this is not a standard error for these functions.
+ *
+ * Note:
+ * Some items like the DropDownList and ListView need further initialization during WM_INITDIALOG
+ * as some of their settings are not exposed as style flags, but need setting using
+ * SendDlgItemMessage().
+ * Use a WM_INITDIALOG callback and perform the init steps in that.
+ */
+
+/*
+ * Undefine some Windows defines to make sure calling these functions sends the call to the DlgBase
+ * implementation.
+ */
+#undef SendMessage
+#undef SendDlgItemMessage
+
+class CallbackBase;
+
+class DlgBase
+{
+protected:
+    enum class DlgMode
+    {
+        PROMPT,
+        INDIRECT,
+    };
+
+    DlgBase(const std::wstring& caption, SHORT x, SHORT y, SHORT w, SHORT h);
+    ~DlgBase();
+
+    //void AddUpDownControl(DWORD id, SHORT x, SHORT y, SHORT w, SHORT h);
+    //void AddDropDownList(DWORD id, SHORT x, SHORT y, SHORT w, SHORT drop_distance);
+    //void AddListView(DWORD id,
+    //                 SHORT x, SHORT y,
+    //                 SHORT w, SHORT h,
+    //                 bool single_selection, bool owner_data);
+    //void AddIPEditControl(DWORD id, SHORT x, SHORT y);
+
+    INT_PTR SpawnDialogBox(const DlgBase* parent, const DlgMode mode);
+
+    LRESULT SendMessage(UINT msg, WPARAM wparam, LPARAM lparam);
+    BOOL ShowDialogBox(int show_window_cmd);
+    BOOL UpdateWindow();
+    void SetReturnCode(INT_PTR return_code);
+    BOOL DestroyDialog();
+
+    void RegisterCloseEventCallback(std::function<bool()> cb);
+
+private:
+    LRESULT SendDlgItemMessage(int item_id, UINT msg, WPARAM wparam, LPARAM lparam);
+    bool SetFocus(DWORD id);
+
+    DWORD GetNextID();
+    SIZE_T AddObject(const std::vector<BYTE>& object);
+    void SetNewStyle(SIZE_T obj_offset, DWORD ex_style, DWORD style);
+
+    void RegisterWmControlCallback(DWORD id, std::function<bool(WORD)> cb);
+
+    bool DestroyCallback();
+    bool NcDestroyCallback();
+
+    static DWORD ms_ref_count;
+
+    std::vector<BYTE> m_window;
+    HWND m_handle;
+    DlgMode m_mode;
+    bool m_return_code_set;
+    INT_PTR m_return_code;
+
+    bool SetMenuBar();
+    void AppendMenuItems(const MenuData* item);
+    std::vector<BYTE> m_menu;
+    MenuData m_menu_children;

```

Comment:
I think `m_menu_data` or something would make more sense. Especially since you're doing `m_menu_children.m_children`.